src/devices/src/virtio/queue.rs

```diff
@@ -393,6 +427,82 @@ impl Queue {
         let addr = self.avail_ring.unchecked_add(2);
         Wrapping(mem.read_obj::<u16>(addr).unwrap())
     }
+
+    /// Get the value of the used event field of the avail ring.
+    #[inline(always)]
+    pub fn used_event(&self, mem: &GuestMemoryMmap) -> Wrapping<u16> {
+        // We need to find the `used_event` field from the avail ring.
+        let used_event_addr = self
+            .avail_ring
+            .unchecked_add(u64::from(4 + 2 * self.actual_size()));
+
+        Wrapping(mem.read_obj::<u16>(used_event_addr).unwrap())
+    }
+
+    /// Helper method that writes `val` to the `avail_event` field of the used ring.
+    fn set_avail_event(&mut self, val: u16, mem: &GuestMemoryMmap) {
+        let avail_event_addr = self
+            .used_ring
+            .unchecked_add(u64::from(4 + 8 * self.actual_size()));
+
+        mem.write_obj(val, avail_event_addr).unwrap();
+    }
+
+    /// Try to enable notification events from the guest driver. Returns true if notifications were
+    /// successfully enabled. Otherwise it means that one or more descriptors can still be consumed
+    /// from the available ring and we can't guarantee that there will be a notification. In this
+    /// case the caller might want to consume the mentioned descriptors and call this method again.
+    pub fn try_enable_notification(&mut self, mem: &GuestMemoryMmap) -> bool {
+        // If the device doesn't use notification suppression, we'll continue to get notifications
+        // no matter what.
+        if !self.uses_notif_suppression {
+            return true;
+        }
+
+        if self.len(mem) != 0 {
+            return false;
+        }
+
+        // Set the next expected avail_idx as avail_event.
+        self.set_avail_event(self.next_avail.0, mem);
+
+        // Make sure all subsequent reads are performed after `set_avail_event`.
+        fence(Ordering::SeqCst);
+
+        // If the actual avail_idx is different than next_avail one or more descriptors can still
+        // be consumed from the available ring.
+        self.next_avail.0 == self.avail_idx(mem).0
+    }
+
+    /// Enable/disable notification suppression.
+    pub fn toggle_notif_suppression(&mut self, enabled: bool) {
+        self.uses_notif_suppression = enabled;
+    }
+
+    /// Check if we need to kick the guest.
+    ///
+    /// Please note this method has side effects: once it returns `true`, it considers the
+    /// driver will actually be notified, and won't return `true` again until the driver
+    /// updates `used_event` and/or the notification conditions hold once more.
+    ///
+    /// This is similar to the `vring_need_event()` method implemented by the Linux kernel.
+    pub fn prepare_kick(&mut self, mem: &GuestMemoryMmap) -> bool {
```

Comment:
Nit: should we rename to `needs_kick` ?

Reply:
IMO `needs_kick` might imply that we are performing only a check, but we are actually also changing the state here. That's why I named it `prepare_kick`. But if you have strong preferences I can rename it.